google/cloud/forseti/scanner/scanners/external_project_access_scanner.py

```diff
@@ -0,0 +1,199 @@
+        """
+        for violation in violations:
+
+            violation_data = {
+                'full_name': violation.full_name,
+                'member': violation.member,
+                'rule_ancestor': violation.rule_ancestor.name
+            }
+
+            yield {
+                'resource_id': violation.resource_id,
+                'resource_type': violation.resource_type,
+                'full_name': violation.full_name,
+                'rule_index': violation.rule_index,
+                'rule_name': violation.rule_name,
+                'violation_type': violation.violation_type,
+                'violation_data': violation_data,
+                'resource_data': violation.resource_data
+            }
+
+    def _retrieve(self):
+        # pylint: disable=line-too-long
+        """Retrieve the project ancestries for all users.
+
+        Returns:
+            dict: User project relationship.
+            {"user1@example.com": [[Project("1234"), Organization("1234567")],
+                                  [Project("12345"), Folder("ABCDEFG"), Organization("1234567")]],
+             "user2@example.com": [[Project("1234"), Organization("1234567")],
+                                  [Project("12345"), Folder("ABCDEFG"), Organization("1234567")]]}
+        """
+        # pylint: enable=line-too-long
+        # This dictionary is the result of the scan.  The key
+        # is the user ID.  The value is a list of lists of ancestries.
+        user_to_project_ancestries_map = {}
+        user_count = 0
+
+        start_time = time.time()
+
+        user_emails = storage_helpers.get_user_emails(self.service_config)
+
+        for user_email in user_emails:
+            user_count += 1
+
+            user_scoped_credential = \
+                api_helpers.get_delegated_credential(user_email, SCOPES)
+
+            client = CloudResourceManagerClient(
+                global_configs=self.inventory_configs,
+                credentials=user_scoped_credential)
+
+            try:
+                project_response = client.get_projects()
+
+                project_ids = \
+                    api_data_helpers.extract_project_ids(project_response)
+                for project_id in project_ids:
+                    if project_id not in DISCOVERED_ANCESTRIES.keys():
+                        ancestries = \
+                            client.get_project_ancestry(project_id)
+                        ancestry_resources = \
+                            api_data_helpers.cast_to_gcp_resources(
+                                ancestries)
+                        DISCOVERED_ANCESTRIES[project_id] = ancestry_resources
+
+                    if user_email not in user_to_project_ancestries_map.keys():
+                        user_to_project_ancestries_map[user_email] = []
+
+                    user_to_project_ancestries_map[user_email].append(
+                        DISCOVERED_ANCESTRIES[project_id])
+            except KeyError:
```

Comment:
How would you get a key error here, since you are explicitly setting the key in the if: block above?

Also, this try block is far too large, you could be catching completely unrelated keyerrors if there is a bug or issue elsewhere in this code.